### PR TITLE
feat(api): auth support for rds models

### PR DESCRIPTION
**Pull request by sundersc**
#### Description of changes

Prep work for supporting auth directive on model level.
- Uses AppSync util methods for authorization.
- Not tested E2E as we are still waiting for AppSync util methods to be live.

##### CDK / CloudFormation Parameters Changed

NA

#### Issue #, if available

NA

#### Description of how you validated changes
- Snapshot testing

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] Any CDK or CloudFormation parameter changes are called out explicitly

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts [hunk 1]
```diff
@@ -7,3 +9,178 @@ export const generateDefaultRDSExpression = (): string => {
...
+  });
+  return expressions;
+};
+
+const convertAuthRoleToVtl = (role: RoleDefinition, fields: string[]): Expression => {
+  const allowedFields = getAllowedFields(role, fields).map((field) => str(field));
+
+  // Api Key
+  if (role.provider === 'apiKey') {
+    return qref(
```
Comment (phani-srikar) [inline packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts:31]: There is too much complexity in this method, can we break this down based on the provider and/or strategy?

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/mutation.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        ref('authRules'),
+        str(operation),
+        ref('ctx.args.input'),
+        ref('ctx.source'),
+      ),
+    ),
+  );
+  expressions.push(validateAuthResult());
+  return printBlock('Authorization rules')(compoundExpression(expressions));
+};
```
Comment (phani-srikar) [inline packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/mutation.ts:80]: The above 3 methods `generateAuthExpressionFor...` has significant parts in common, can centralize these in a single method.

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/query.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+export const generateAuthExpressionForQueries = (
+  ctx: TransformerContextProvider,
+  providers: ConfiguredAuthProviders,
+  roles: Array<RoleDefinition>,
+  fields: ReadonlyArray<FieldDefinitionNode>,
+  def: ObjectTypeDefinitionNode,
+  indexName: string | undefined,
+): string => {
+  const expressions = [];
```
Comment (phani-srikar) [inline packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/query.ts:15]: Most of these inputs are not being used in the method, can you remove unused ones?

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/subscription.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { and, compoundExpression, ifElse, iff, list, methodCall, not, obj, printBlock, ref, set } from 'graphql-mapping-template';
+import { ConfiguredAuthProviders, RoleDefinition } from '../../../utils';
+import { generateAuthRulesFromRoles, validateAuthResult } from './common';
+
+export const generateAuthExpressionForSubscriptions = (providers: ConfiguredAuthProviders, roles: Array<RoleDefinition>): string => {
+  const expressions = [];
```
Comment (phani-srikar) [inline packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/subscription.ts:6]: `providers` input is not used in this method, check and remove.

File: packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-subscription-auth.test.ts.snap [hunk 5]
```diff
@@ -0,0 +1,2203 @@
...
+  $util.unauthorized()
+#end
+$util.toJson({})
+## [End] Sandbox Mode Disabled. **"
+`;
+
+exports[`Verify RDS Model level Auth rules on subscriptions: should successfully transform cognito auth rules 30`] = `
+"## [Start] Subscription Response template. **
+#if( !$util.isNullOrEmpty($ctx.args.filter) )
+$extensions.setSubscriptionFilter($util.transform.toSubscriptionFilter($ctx.args.filter))
```
Comment (phani-srikar) [inline packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-subscription-auth.test.ts.snap:583]: nit: general comment for the VTL generated in multiple of these snapshots is that the indentation is off.

**Review by phani-srikar — CHANGES_REQUESTED** [review]

**Review by phani-srikar — CHANGES_REQUESTED** [review]
There are some unit test failures that need to be fixed.

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/subscription.ts (hunk 4 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/subscription.ts:6]: Same as above. These got added here to implement the VTL Generator interface.

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/query.ts (hunk 3 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/query.ts:16]: DDB and RDS VTL generator implement a common interface, most of these arguments are used in DDB but not in RDS. So, there is going to be unused arguments somewhere. We can avoid this by creating datasource specific config types (may be in a separate PR)

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/mutation.ts (hunk 2 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/mutation.ts]: This has been refactored into a helper method.

File: packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-subscription-auth.test.ts.snap (hunk 5 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-subscription-auth.test.ts.snap:583]: These are generated by mapping-template helpers and jest. Not much we can do in the scope of this PR.

File: packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-mutation-auth.test.ts.snap [hunk 6]
```diff
@@ -0,0 +1,1995 @@
...
+  \\"type\\": \\"public\\",
+  \\"provider\\": \\"apiKey\\",
+  \\"allowedFields\\":   [\\"id\\", \\"title\\"]
+}))
+#set( $authResult = $util.authRules.mutationAuth($authRules, \\"create\\", $ctx.args.input) )
+#if( !$authResult || ($authResult && !$authResult.authorized) )
+$util.unauthorized
+#end
+#if( $authResult && !$util.isNullOrEmpty($authResult.authorizedInput) )
+  #set( $ctx.args.input = $authResult.authorizedInput )
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/rds-mutation-auth.test.ts.snap:17]: I understand this is not a desired behavior if we are changing anything related to the identity. But authorizing the input and using it all the way in the resolvers seems to be okay to me, we do this currently for subscription filters too. For any reason, if we need the original input in the future, we can assign it to a stash variable.

File: packages/amplify-graphql-sql-transformer/src/graphql-sql-transformer.ts [hunk 7]
```diff
@@ -105,13 +105,32 @@ export class SqlTransformer extends TransformerPluginBase {
           ),
         );
 
+        resolver.addToSlot(
+          'postAuth',
+          MappingTemplate.s3MappingTemplateFromString(
+            generateAuthExpressionForSandboxMode(context.transformParameters.sandboxModeEnabled),
+            `${config.resolverTypeName}.${config.resolverFieldName}.{slotName}.{slotIndex}.req.vtl`,
+          ),
+        );
```
Comment (phani-srikar) [inline packages/amplify-graphql-sql-transformer/src/graphql-sql-transformer.ts:114]: Having atleast a unit test for these resolver changes would help.

**Review by phani-srikar — APPROVED** [review]
Approving pending E2E testing to parallelize that effort.

File: packages/amplify-graphql-sql-transformer/src/graphql-sql-transformer.ts (hunk 7 above)
Comment (phani-srikar) [reply packages/amplify-graphql-sql-transformer/src/graphql-sql-transformer.ts:114]: Other alternative is to pull these changes into another sub-feature branch in the repo and merge post all E2Es pass.

**Review by palpatim — APPROVED** [review]

**Pull request merged**